src/components/Exercise.tsx

```diff
@@ -35,37 +35,48 @@ export interface ExerciseProps {
   onAnswerChange: () => void;
   onAnswerSave: () => void;
   onNextStep: () => void;
-  canUpdateCurrentStep: boolean;
-  attempt_number: number;
   apiIsPending: boolean;
-  available_points: AvailablePoints;
+}
+
+export interface ExerciseWithStepDataProps extends ExerciseBaseProps {
+  step: StepWithData;
+}
+
+export interface ExerciseWithQuestionStatesProps extends ExerciseBaseProps {
+  questionStates: { [key: ID]: QuestionState };
 }
 
 export const Exercise = ({
   numberOfQuestions, questionNumber, step, exercise, canAnswer, needsSaved, ...props
-}: ExerciseProps) => (
+}: ExerciseWithStepDataProps | ExerciseWithQuestionStatesProps) => (
   <StyledTaskStepCard
     step={step}
     questionNumber={questionNumber}
     numberOfQuestions={numberOfQuestions}
   >
     <Preamble exercise={exercise} />
 
-    {exercise.questions.map((q) =>
-      <ExerciseQuestion
-        {...props}
-        {...step}
-        exercise_uid={exercise.uid}
-        key={q.id}
-        question={q}
-        questionNumber={questionNumber}
-        choicesEnabled={canAnswer}
-        canAnswer={canAnswer}
-        needsSaved={needsSaved}
-        canUpdateCurrentStep={canAnswer}
-        displaySolution={true}
-        answerId={step.answer_id}
-      />
+    {exercise.questions.map((q) => {
+      const state = props['questionStates'] && props['questionStates'][q.id];
+      const detailedSolution = step['detailedSolution'] || (state['solution'] && state['solution']['content_html'])
+      return (
+        <ExerciseQuestion
+          {...props}
+          {...step}
+          {...state}
```

Comment:
maybe this would help
`{...('feedbackHtml' in step ? step : {...step, ...props['questionStates'][q.id])}}`